extensions/amp-sidebar/0.1/amp-sidebar.js

```diff
@@ -0,0 +1,254 @@
+import {setStyles} from '../../../src/style';
+
+
+/** @const */
+const EXPERIMENT = 'amp-sidebar';
+
+/** @const */
+const TAG = 'AmpSidebar';
+
+/** @const */
+const WHITELIST_ = ['AMP-ACCORDION', 'AMP-FIT-TEXT', 'AMP-IMG'];
+
+/** @const */
+const IOS_SAFARI_BOTTOMBAR_HEIGHT = '10vh';
+
+export class AmpSidebar extends AMP.BaseElement {
+
+  /** @override */
+  isLayoutSupported(layout) {
+    return layout == Layout.CONTAINER;
+  }
+
+  /** @override */
+  isReadyToBuild() {
+    return false;
+  }
+
+  /** @override */
+  buildCallback() {
+    /** @const @private {boolean} */
+    this.isExperimentOn_ = isExperimentOn(this.getWin(), EXPERIMENT);
+
+    /** @private @const {!Window} */
+    this.win_ = this.getWin();
+
+    /** @private @const {!Document} */
+    this.document_ = this.win_.document;
+
+    /** @private @const {!Element} */
+    this.documentElement_ = this.document_.documentElement;
+
+    /** @private @const {string} */
+    this.direction_ = this.element.getAttribute('direction');
+
+    /** @private @const {!Viewport} */
+    this.viewport_ = this.getViewport();
+
+    /** @private @const {boolean} */
+    this.hasMask_ = false;
+
+    /** @private @const {boolean} */
+    this.isIosSafari_ = platform.isIos() && platform.isSafari();
+
+    if (this.direction_ != 'left' && this.direction_ != 'right') {
+      const pageDir = this.document_.body.getAttribute('dir') || 'ltr';
+      this.direction_ = (pageDir == 'rtl') ? 'right' : 'left';
+      this.element.setAttribute('direction', this.direction_);
+    }
+
+    if (!this.isExperimentOn_) {
+      dev.warn(TAG, `Experiment ${EXPERIMENT} disabled`);
+      return;
+    }
+    if (!this.checkWhitelist_()) {
+      return;
+    }
+
+    if (this.isIosSafari_) {
+      this.fixIosElasticScrollLeak_();
+    }
+
+    this.adjustPadding_();
+
+    if (this.documentElement_.classList.contains('amp-sidebar-open')) {
+      // Create the mask if the sidebar is rendered in open mode.
+      this.createMask_();
+    }
+
+    this.documentElement_.addEventListener('keydown', event => {
+      // Close sidebar on ESC.
+      if (event.keyCode == 27) {
+        this.close_();
+      }
+    });
+    //TODO (skrish, #2712) Add history support on back button.
+    this.registerAction('activate', this.open_.bind(this));
+    this.registerAction('toggle', this.toggle_.bind(this));
+    this.registerAction('open', this.open_.bind(this));
+    this.registerAction('close', this.close_.bind(this));
+  }
+
+  /**
+   * @private
+   */
+  adjustPadding_() {
+    const viewerPaddingTop = this.viewport_.getPaddingTop();
+    if (viewerPaddingTop) {
+      // viewerPaddingTop exists when AMP page is rendered inside search
+      // carousel.
+      const div = this.document_.createElement('div');
+      setStyles(div, {
+        'height': viewerPaddingTop + 'px',
+        'width': '100%',
+      });
+      const firstChild = this.element.firstChild;
+      this.element.insertBefore(div, firstChild);
```

Comment:
This is hard to guarantee to work - this will skew many possible design situations. The best way is still to adjust `top` position. If for some reason this is not good - we can explore updating padding top. But we need to have relative position's reference point to be in the right place.